src/cdk/scrolling/virtual-for-of.ts

```diff
@@ -0,0 +1,310 @@
+          pairwise(),
+          // Use `_changeDataSource` to disconnect from the previous data source and connect to the
+          // new one, passing back a stream of data changes which we run through `switchMap` to give
+          // us a data stream that emits the latest data from whatever the current `DataSource` is.
+          switchMap(([prev, cur]) => this._changeDataSource(prev, cur)),
+          // Replay the last emitted data when someone subscribes.
+          shareReplay(1));
+
+  private _differ: IterableDiffer<T> | null = null;
+
+  private _data: T[];
+
+  private _renderedItems: T[];
+
+  private _renderedRange: Range;
+
+  private _templateCache: EmbeddedViewRef<CdkVirtualForOfContext<T>>[] = [];
+
+  private _needsUpdate = false;
+
+  constructor(
+      private _viewContainerRef: ViewContainerRef,
+      private _template: TemplateRef<CdkVirtualForOfContext<T>>,
+      private _differs: IterableDiffers,
+      @Host() private _viewport: CdkVirtualScrollViewport) {
+    this.dataStream.subscribe(data => this._data = data);
+    this._viewport.renderedRangeStream.subscribe(range => this._onRenderedRangeChange(range));
+    this._viewport.connect(this);
+  }
+
+  /**
+   * Get the client rect for the given index.
+   * @param index The index of the data element whose client rect we want to measure.
+   * @return The combined client rect for all DOM elements rendered as part of the given index.
+   *     Or null if no DOM elements are rendered for the given index.
+   * @throws If the given index is not in the rendered range.
+   */
+  measureClientRect(index: number): ClientRect | null {
+    if (index < this._renderedRange.start || index >= this._renderedRange.end) {
+      throw Error('Error: attempted to measure an element that isn\'t rendered.');
+    }
+    index -= this._renderedRange.start;
+    let view = this._viewContainerRef.get(index) as
+        EmbeddedViewRef<CdkVirtualForOfContext<T>> | null;
+    if (view && view.rootNodes.length) {
+      let minTop = Infinity;
+      let minLeft = Infinity;
+      let maxBottom = -Infinity;
+      let maxRight = -Infinity;
+
+      // There may be multiple root DOM elements for a single data element, so we merge their rects.
+      for (let i = 0, ilen = view.rootNodes.length; i < ilen; i++) {
+        let rect = (view.rootNodes[i] as Element).getBoundingClientRect();
+        minTop = Math.min(minTop, rect.top);
+        minLeft = Math.min(minLeft, rect.left);
+        maxBottom = Math.max(maxBottom, rect.bottom);
+        maxRight = Math.max(maxRight, rect.right);
+      }
+
+      return {
+        top: minTop,
+        left: minLeft,
+        bottom: maxBottom,
+        right: maxRight,
+        height: maxBottom - minTop,
+        width: maxRight - minLeft
+      };
+    }
+    return null;
+  }
+
+  ngDoCheck() {
+    if (this._differ && this._needsUpdate) {
+      const changes = this._differ.diff(this._renderedItems);
```

Comment:
This is really pricey, is there a more reactive way we can do this updating?

Reply:
I think we could differentiate between needing an update due to scrolling and rendering a new portion of the list (can do something simpler) vs needing an update due to the underlying data changing (in which case I think we'd have to do this diff). I added a TODO to add some logic for that. Also added a TODO to calculate the minimum set of remove & inserts since I'm currently just removing all of the views and re-inserting them in the right order